Source/Core/InputCommon/ControllerInterface/Win32/Win32.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "InputCommon/ControllerInterface/Win32/Win32.h"
+
+#include <windows.h>
+
+#include <thread>
+
+#include "Common/Flag.h"
+#include "Common/Logging/Log.h"
+#include "Common/ScopeGuard.h"
+#include "InputCommon/ControllerInterface/DInput/DInput.h"
+#include "InputCommon/ControllerInterface/XInput/XInput.h"
+
+constexpr UINT WM_DOLPHIN_STOP = WM_USER;
+
+static HWND s_hwnd;
+static HWND s_message_window;
+static std::thread s_thread;
+
+static LRESULT CALLBACK WindowProc(HWND hwnd, UINT message, WPARAM wparam, LPARAM lparam)
+{
+  if (message == WM_INPUT_DEVICE_CHANGE)
+  {
+    ciface::DInput::PopulateDevices(s_hwnd);
+    ciface::XInput::PopulateDevices();
+    g_controller_interface.InvokeHotplugCallbacks();
+  }
+
+  return DefWindowProc(hwnd, message, wparam, lparam);
+}
+
+void ciface::Win32::Init(void* hwnd)
+{
+  s_hwnd = static_cast<HWND>(hwnd);
+  XInput::Init();
+
+  s_thread = std::thread([] {
+    if (FAILED(CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED)))
+    {
+      ERROR_LOG(SERIALINTERFACE, "CoInitializeEx failed: %i", GetLastError());
+      return;
+    }
+    Common::ScopeGuard uninit([] { CoUninitialize(); });
```

Comment:
Note that this will uninitialize the entire COM library. If you have other COM modules you're in for trouble.

Reply:
The [docs](https://msdn.microsoft.com/en-us/library/windows/desktop/ms688715(v=vs.85).aspx) say:

> Closes the COM library on the current thread [...] A thread must call CoUninitialize once for each successful call it has made to the CoInitialize or CoInitializeEx function

Is there something else it should do instead?

Reply:
I had completely missed it was a thread, sorry.

Reply:
Cool, no worries!